crates/networking/p2p/rlpx/l2/l2_connection.rs

```diff
@@ -0,0 +1,934 @@
+    let l2_state = established.l2_state.connection_state_mut()?;
+    if !established.blockchain.is_synced() {
+        debug!("Not processing new block, blockchain is not synced");
+        return Ok(false);
+    }
+    if l2_state
+        .store_rollup
+        .contains_batch(&msg.batch.number)
+        .await?
+    {
+        debug!("Batch {} already sealed, ignoring it", msg.batch.number);
+        return Ok(false);
+    }
+    if msg.batch.first_block == msg.batch.last_block {
+        // is empty batch
+        return Ok(false);
+    }
+    if l2_state.latest_block_added < msg.batch.last_block {
+        debug!(
+            "Not processing batch {} because the last block {} is not added yet",
+            msg.batch.number, msg.batch.last_block
+        );
+        return Ok(false);
+    }
+
+    let hash = batch_hash(&msg.batch);
+
+    let recovered_lead_sequencer =
+        get_pub_key(msg.recovery_id, &msg.signature, hash.0).map_err(|e| {
+            log_peer_error(
+                &established.node,
+                &format!("Failed to recover lead sequencer: {e}"),
+            );
+            RLPxError::CryptographyError(e.to_string())
+        })?;
+
+    if !validate_signature(recovered_lead_sequencer) {
+        return Ok(false);
+    }
+
+    let mut signature = [0u8; 68];
+    signature[..64].copy_from_slice(&msg.signature[..]);
+    signature[64..].copy_from_slice(&msg.recovery_id[..]);
+    l2_state
+        .store_rollup
+        .store_signature_by_batch(msg.batch.number, signature)
+        .await?;
+    Ok(true)
+}
+
+async fn process_new_block(established: &mut Established, msg: &NewBlock) -> Result<(), RLPxError> {
+    let l2_state = established.l2_state.connection_state_mut()?;
+    l2_state
+        .blocks_on_queue
+        .entry(msg.block.header.number)
+        .or_insert_with(|| msg.block.clone());
+
+    let mut next_block_to_add = l2_state.latest_block_added + 1;
+    while let Some(block) = l2_state.blocks_on_queue.remove(&next_block_to_add) {
+        // This check is necessary if a connection to another peer already applied the block but this connection
+        // did not register that update.
+        if let Ok(Some(_)) = established.storage.get_block_body(next_block_to_add).await {
+            l2_state.latest_block_added = next_block_to_add;
+            next_block_to_add += 1;
+            continue;
+        }
+        established
+            .blockchain
+            .add_block(&block)
+            .await
+            .inspect_err(|e| {
+                log_peer_error(
+                    &established.node,
+                    &format!(
+                        "Error adding new block {} with hash {:?}, error: {e}",
+                        block.header.number,
+                        block.hash()
+                    ),
+                );
+            })?;
+        let block_hash = block.hash();
+
+        apply_fork_choice(&established.storage, block_hash, block_hash, block_hash)
```

Comment:
This probably requires some debate, but I think we should consider applying fork choice when sealing the batch to avoid some edge cases where the block is on state but is orphaned in terms of batch.

Reply:
Added an issue for this: https://github.com/lambdaclass/ethrex/issues/3597